Add platform-specific handling for socket operations

vitimiti opened the pull request:
Introduced conditional compilation to differentiate behavior between Windows (_WIN32) and other platforms.

Updated functions to handle type and method discrepancies for compatibility.

Included necessary headers and definitions for non-Windows environments.

vitimiti committed:
Add platform-specific handling for socket operations

Introduced conditional compilation to differentiate behavior between
Windows (_WIN32) and other platforms.

Updated functions to handle type and method discrepancies for
compatibility.

Included necessary headers and definitions for non-Windows environments.

Signed-off-by: Victor Matia <vmatir@outlook.com>

xezon commented on Lzhl_tcp.cpp:
```diff
@@ -54,7 +54,11 @@ SOCKET lzhl_socket( int af, int type, int protocol )
 
 SOCKET lzhl_accept( SOCKET s, struct sockaddr* addr, int* addrlen )
```
I suggest add a note above this function that we have added the non-windows stuff

`// TheSuperHackers @feature vitimiti 13/03/2024 Your text here`

xezon commented on Lzhl_tcp.cpp:
```diff
@@ -252,5 +260,10 @@ int lzhl_closesocket( SOCKET sock )
+#if _WIN32
 	return closesocket( sock );
+#else
+    return 0;
```
Nothing leaks here?

vitimiti replied:
I misread the documentation about UNIX sockets closing. Let me fix that and add the comment :)

vitimiti committed:
Add UNIX sockets support in Lzhl_tcp

Added missing UNIX socket closing and recommended comments.

Signed-off-by: Victor Matia <vmatir@outlook.com>

xezon commented on Lzhl_tcp.cpp:
```diff
@@ -212,7 +218,14 @@ int lzhl_recv( SOCKET sock, char* buf, int bufSz, int flags )
 			ls.dDisp = 0;
 			ls.dBuf = new BYTE[ dataSz ];
+
+            // TheSuperHackers @feature vitimiti 13/03/2025 Added UNIX sockets support
```
One comment above the entire function is enough. Makes it cleaner.

vitimiti replied:
One comment at the top for each function then? Do I understand correctly?

xezon replied:
As far as I can see you only touched `lzhl_accept` function. So one comment above that totally suffices.

vitimiti replied:
I touched `lzhl_recv` and `lzhl_closesocket`, as well as the header `Lzhl_tcp.h`

xezon replied:
Ah ok my mistake. I somehow thought this is all in one function with the way github shows this. Your comments look fine then.

xezon commented on Lzhl_tcp.cpp:
```diff
@@ -54,7 +54,13 @@ SOCKET lzhl_socket( int af, int type, int protocol )
+#if _WIN32
 	SOCKET sock = accept( s, addr, addrlen );
+#else
+	SOCKET sock = accept( s, addr, (socklen_t*)addrlen );
```
On second look, is this safe? This is a `reinterpret_cast` and if `socklen_t` is larger than 4 bytes then this is an error.

xezon commented on Lzhl_tcp.cpp:
```diff
@@ -212,7 +218,14 @@ int lzhl_recv( SOCKET sock, char* buf, int bufSz, int flags )
+#if _WIN32
 			int Ok = LZHLDecompress( dh, ls.dBuf, &dataSz, compBuf, &compSz );
+#else
+            int Ok = LZHLDecompress( dh, ls.dBuf, (size_t*)(&dataSz), compBuf, (size_t*)(&compSz) );
```
This is definitely an unsafe. `dataSz` is int 4 bytes but `size_t` could be 8 bytes.

vitimiti replied:
The function is designed with `size_t`, but fair point, I will add #if blocks for UNIX

xezon replied:
They are declared as `unsigned int dataSz, compSz;` 4 bytes.

vitimiti replied:
I am working on a fix now. I know the variables are declared as such, but they are incompatible with Linux with how the `LZHLDecompress` function is defined.

xezon replied:
The easiest way to fix is 

```cpp
size_t dataSz64 = dataSz;
size_t compSz64 = compSz;
int Ok = LZHLDecompress( dh, ls.dBuf, &dataSz64, compBuf, &compSz64 );
```

vitimiti replied:
I have done, but added asserts to ensure the resulting values don't have bits dropped when converting them back

vitimiti committed:
Added UNIX specific safety

Added safety to UNIX specific casts.

Signed-off-by: Victor Matia <vmatir@outlook.com>

vitimiti commented:
Have a look at the latest push, I can't think of a better way to assert safety between the original types and the types expected by GCC/Clang on Linux, unless you have a better idea that I can implement, I think this is safer.

xezon commented on Lzhl_tcp.cpp:
```diff
@@ -212,7 +223,19 @@ int lzhl_recv( SOCKET sock, char* buf, int bufSz, int flags )
+            assert( dataSzLongUint <= UINT_MAX );
+            assert( compSzLongUint <= UINT_MAX );
+            dataSz = (unsigned int)dataSzLongUint;
+            compSz = (unsigned int)compSzLongUint;
```
It looks like `dataSz`, `compSz` are not used after this point so you can skip the assignment and assert here.

xezon commented on Lzhl_tcp.cpp:
```diff
@@ -153,6 +163,7 @@ int lzhl_send( SOCKET sock, const char* data, int dataSz, int flags )
 
 int lzhl_recv( SOCKET sock, char* buf, int bufSz, int flags )
 {
+    // TheSuperHackers @feature vitimiti 13/03/2025 Added UNIX sockets support
```
I think the way you had the comment closer to your change made more sense before. My mistake.

vitimiti replied:
It's okay, I'll update it when I get home, 30-60 minutes

vitimiti replied:
Fixed all back, and removed the unnecessary asserts and assignments

xezon reviewed (commented):
This looks like it should work.

vitimiti committed:
Changed comments back and removed unused code

Removed code unused for asserts and assignments and moved comments back
as requested.

Signed-off-by: Victor Matia <vmatir@outlook.com>

xezon reviewed (approved):
Looks correct to me.
